src/node_contextify.cc

```diff
@@ -1391,13 +1395,268 @@ void MicrotaskQueueWrap::RegisterExternalReferences(
   registry->Register(New);
 }
 
+Local<FunctionTemplate> NodeRealm::GetConstructorTemplate(
+    IsolateData* isolate_data) {
+  Local<FunctionTemplate> tmpl =
+      isolate_data->node_realm_constructor_template();
+  if (tmpl.IsEmpty()) {
+    Isolate* isolate = isolate_data->isolate();
+    tmpl = NewFunctionTemplate(isolate, New);
+    tmpl->Inherit(AsyncWrap::GetConstructorTemplate(isolate_data));
+    tmpl->InstanceTemplate()->SetInternalFieldCount(1);
+
+    SetProtoMethod(isolate, tmpl, "start", Start);
+    SetProtoMethod(isolate, tmpl, "load", Load);
+    SetProtoMethod(isolate, tmpl, "stop", Stop);
+    SetProtoMethod(isolate, tmpl, "signalStop", SignalStop);
+    SetProtoMethod(isolate, tmpl, "tryCloseAllHandles", TryCloseAllHandles);
+    SetProtoMethod(isolate, tmpl, "internalRequire", InternalRequire);
+
+    isolate_data->set_node_realm_constructor_template(tmpl);
+  }
+  return tmpl;
+}
+
+void NodeRealm::CreatePerIsolateProperties(
+    IsolateData* isolate_data, v8::Local<v8::ObjectTemplate> target) {
+  SetConstructorFunction(isolate_data->isolate(),
+                         target,
+                         "NodeRealm",
+                         GetConstructorTemplate(isolate_data),
+                         SetConstructorFunctionFlag::NONE);
+}
+
+void NodeRealm::RegisterExternalReferences(
+    ExternalReferenceRegistry* registry) {
+  registry->Register(New);
+  registry->Register(Start);
+  registry->Register(Load);
+  registry->Register(SignalStop);
+  registry->Register(Stop);
+  registry->Register(TryCloseAllHandles);
+  registry->Register(InternalRequire);
+}
+
+NodeRealm::NodeRealmScope::NodeRealmScope(NodeRealm* w)
+    : EscapableHandleScope(w->isolate_),
+      Scope(w->context()),
+      Isolate::SafeForTerminationScope(w->isolate_),
+      w_(w),
+      orig_can_be_terminated_(w->can_be_terminated_) {
+  w_->can_be_terminated_ = true;
+}
+
+NodeRealm::NodeRealmScope::~NodeRealmScope() {
+  w_->can_be_terminated_ = orig_can_be_terminated_;
+}
+
+Local<Context> NodeRealm::context() const {
+  return context_.Get(isolate_);
+}
+
+NodeRealm::NodeRealm(Environment* env, Local<Object> object)
+    : isolate_(env->isolate()), wrap_(env->isolate(), object) {
+  AddEnvironmentCleanupHook(env->isolate(), CleanupHook, this);
+  object->SetAlignedPointerInInternalField(0, this);
+
+  Local<Context> outer_context = env->context();
+  outer_context_.Reset(env->isolate(), outer_context);
+}
+
+NodeRealm* NodeRealm::Unwrap(const FunctionCallbackInfo<Value>& args) {
```

Comment:
Should NodeRealm inherit from BaseObject? This seems pretty similar to BaseObject's work.

Reply:
Yes. Only reason I didn't do this in synchronous-worker is because it was built outside of Node.js core.